File: mobile/app/src/main/java/com/aguiabranca/app/core/domain/badge/BadgeEvaluator.kt
package com.aguiabranca.app.core.domain.badge

import com.aguiabranca.app.core.domain.model.Idea
import com.aguiabranca.app.core.domain.model.IdeaStatus
import com.aguiabranca.app.core.domain.model.User
import java.util.Calendar

object Badges {
    const val PRIMEIRA_IDEIA = "Primeira Ideia"
    const val ESTRATEGISTA = "Estrategista"
    const val INOVADOR_MES = "Inovador do Mês"
    const val IMPACTO_REAL = "Impacto Real"
    const val VISIONARIO = "Visionário"
}

object BadgeEvaluator {

    fun evaluate(user: User, ideas: List<Idea>): Set<String> {
        val authoredIdeas = ideas.filter { it.authorId == user.id }
        val already = user.badges.toSet()
        val unlocked = mutableSetOf<String>()

        if (authoredIdeas.isNotEmpty() && Badges.PRIMEIRA_IDEIA !in already) {
            unlocked += Badges.PRIMEIRA_IDEIA
        }

        if (Badges.ESTRATEGISTA !in already &&
            authoredIdeas.any { it.guidelineId != null && it.status in approvedOrLater }
        ) {
            unlocked += Badges.ESTRATEGISTA
        }

        if (Badges.INOVADOR_MES !in already) {
            val groupedByMonth = authoredIdeas.groupBy { yearMonthOf(it.createdAt) }
            if (groupedByMonth.any { it.value.size >= 5 }) {
                unlocked += Badges.INOVADOR_MES
            }
        }

        if (Badges.IMPACTO_REAL !in already &&
            authoredIdeas.any { it.status == IdeaStatus.IMPLEMENTADA }
        ) {
            unlocked += Badges.IMPACTO_REAL
        }

        if (Badges.VISIONARIO !in already) {
            val distinctGuidelines = authoredIdeas
                .filter { it.guidelineId != null && it.status in approvedOrLater }
                .mapNotNull { it.guidelineId }
                .toSet()
            if (distinctGuidelines.size >= 3) {
                unlocked += Badges.VISIONARIO
            }
        }

        return unlocked
    }

    private val approvedOrLater = setOf(IdeaStatus.APROVADA, IdeaStatus.IMPLEMENTADA)

    private fun yearMonthOf(epochMillis: Long): Int {
        if (epochMillis <= 0) return 0
        val c = Calendar.getInstance().apply { timeInMillis = epochMillis }
        return c.get(Calendar.YEAR) * 100 + c.get(Calendar.MONTH)
    }
}

File: backend/src/AguiaBranca.Domain/Rules/BadgeEvaluator.cs
using AguiaBranca.Domain.Entities;
using AguiaBranca.Domain.Enums;

namespace AguiaBranca.Domain.Rules;

public static class Badges
{
    public const string PrimeiraIdeia = "Primeira Ideia";
    public const string Estrategista = "Estrategista";
    public const string InovadorDoMes = "Inovador do Mês";
    public const string ImpactoReal = "Impacto Real";
    public const string Visionario = "Visionário";

    public static readonly IReadOnlyList<string> All =
        [PrimeiraIdeia, Estrategista, InovadorDoMes, ImpactoReal, Visionario];
}

/// <summary>Porte de <c>BadgeEvaluator.kt</c> (R-06.4). Retorna apenas badges <b>novas</b> (que o usuário ainda não tem).</summary>
public static class BadgeEvaluator
{
    private const int IdeasForInovadorDoMes = 5;
    private const int DistinctGuidelinesForVisionario = 3;

    /// <param name="ideas">Ideias do usuário (as de outros autores são ignoradas).</param>
    /// <param name="monthZone">Fuso usado para agrupar "mês calendário" (padrão: UTC).</param>
    public static IReadOnlySet<string> Evaluate(AppUser user, IEnumerable<Idea> ideas, TimeZoneInfo? monthZone = null)
    {
        monthZone ??= TimeZoneInfo.Utc;
        var authored = ideas.Where(i => i.AuthorId == user.Id).ToList();
        var already = user.Badges.ToHashSet();
        var unlocked = new HashSet<string>();

        if (authored.Count > 0 && !already.Contains(Badges.PrimeiraIdeia))
            unlocked.Add(Badges.PrimeiraIdeia);

        if (!already.Contains(Badges.Estrategista) && authored.Any(i => i.HasStrategicLink && i.IsApprovedOrLater))
            unlocked.Add(Badges.Estrategista);

        if (!already.Contains(Badges.InovadorDoMes) &&
            authored.GroupBy(i => YearMonth(i.CreatedAt, monthZone)).Any(g => g.Count() >= IdeasForInovadorDoMes))
            unlocked.Add(Badges.InovadorDoMes);

        if (!already.Contains(Badges.ImpactoReal) && authored.Any(i => i.Status == IdeaStatus.IMPLEMENTADA))
            unlocked.Add(Badges.ImpactoReal);

        if (!already.Contains(Badges.Visionario))
        {
            var distinct = authored
                .Where(i => i.HasStrategicLink && i.IsApprovedOrLater)
                .Select(i => i.GuidelineId!)
                .Distinct()
                .Count();
            if (distinct >= DistinctGuidelinesForVisionario) unlocked.Add(Badges.Visionario);
        }

        return unlocked;
    }

    private static int YearMonth(DateTime utc, TimeZoneInfo zone)
    {
        var local = TimeZoneInfo.ConvertTimeFromUtc(DateTime.SpecifyKind(utc, DateTimeKind.Utc), zone);
        return local.Year * 100 + local.Month;
    }
}
